mmdet/models/detectors/deformable_detr.py

```diff
@@ -1,12 +1,745 @@
 # Copyright (c) OpenMMLab. All rights reserved.
+import math
+from typing import Dict, Optional, Tuple, Union
+
+import torch
+import torch.nn.functional as F
+from mmcv.cnn.bricks.transformer import (FFN, MultiheadAttention,
+                                         MultiScaleDeformableAttention,
+                                         build_norm_layer)
+from mmengine.model import ModuleList, xavier_init
+from torch import Tensor, nn
+from torch.nn.init import normal_
+
 from mmdet.registry import MODELS
-from .detr import DETR
+from mmdet.structures import OptSampleList
+from mmdet.utils import OptConfigType
+from ..layers import (DetrTransformerDecoder, DetrTransformerDecoderLayer,
+                      DetrTransformerEncoder, DetrTransformerEncoderLayer,
+                      SinePositionalEncoding, inverse_sigmoid)
+from .base_detr import TransformerDetector
 
 
 @MODELS.register_module()
-class DeformableDETR(DETR):
+class DeformableDETR(TransformerDetector):
     r"""Implementation of `Deformable DETR: Deformable Transformers for
-    End-to-End Object Detection <https://arxiv.org/abs/2010.04159>`_"""
+    End-to-End Object Detection <https://arxiv.org/abs/2010.04159>`_
+
+    Code is modified from the `official github repo
+    <https://github.com/fundamentalvision/Deformable-DETR>`_.
+
+    Args:
+        decoder_cfg (:obj:`ConfigDict` or dict): Config of decoder.
+            Defaults to None.
+        bbox_head (:obj:`ConfigDict` or dict): Config for position
+            encoding. Defaults to None.
+        with_box_refine (bool, optional): Whether to refine the references
+            in the decoder. Defaults to `False`.
+        as_two_stage (bool, optional): Whether to generate the proposal
+            from the outputs of encoder. Defaults to `False`.
+        num_feature_levels (int, optional): Number of feature levels.
+            Defaults to 4.
+    """
+
+    def __init__(self,
+                 *args,
+                 decoder_cfg: OptConfigType = None,
+                 bbox_head: OptConfigType = None,
+                 with_box_refine: bool = False,
+                 as_two_stage: bool = False,
+                 num_feature_levels: int = 4,
+                 **kwargs) -> None:
+        self.with_box_refine = with_box_refine
+        self.as_two_stage = as_two_stage
+        self.num_feature_levels = num_feature_levels
+
+        if bbox_head is not None:
+            bbox_head.update(
+                dict(
+                    with_box_refine=with_box_refine,
+                    as_two_stage=as_two_stage,
+                    num_decoder_layers=decoder_cfg['num_layers']))
```

Comment:
Better not to modify the config with hardcode.

Reply:
> Better not to modify the config with hardcode.

@RangiLyu Thanks for your review!
The args have been refacted. The related code is as follows:
```Python
if bbox_head is not None:
    assert 'share_pred_layer' not in bbox_head and \
           'num_pred_layer' not in bbox_head, \
        'The two keyword args `share_pred_layer`, `num_pred_layer` ' \
        'are set in `detector.__init__()`, users should not set ' \
        'them in `bbox_head` config.'
    # The last prediction layer is used to generate proposal
    # from encode feature map when `as_two_stage` is `True`.
    # And all the prediction layers should share parameters
    # when `with_box_refine` is `True`.
    bbox_head['share_pred_layer'] = not with_box_refine
    bbox_head['num_pred_layer'] = (decoder_cfg['num_layers'] + 1) \
        if self.as_two_stage else decoder_cfg['num_layers']
```